### ISSUE TITLE
figure.Figure.autofmt_xdate applied to major xtick labels only

### ISSUE THREAD

**Issue by gitporst** [matplotlib/matplotlib#8128]
**Bug summary**
The `autofmt_xdate` of the `figure.Figure` class by default only formats the major tick labels. Increased functionality including the minor ticks as well would be nice.

**Code for reproduction**
```python
import pandas as pd
import matplotlib.pyplot as plt

fig, ax = plt.subplots()

df = pd.DataFrame(data=[1] * 5,
                  index=pd.date_range(start='2017-01-01', periods=5),
                  columns=['A'])
df.plot(ax=ax)
fig.autofmt_xdate()
plt.show()
```

**Actual outcome**
![autofmt_xdate_issue](https://cloud.githubusercontent.com/assets/20670129/23210053/2c67b024-f8fc-11e6-8e1d-bfb66c7da2ac.png)

**Expected outcome**
It would be nice, if autofmt_xdate would be applied to the minor labels as well. This could be achieved, for example, by adding the keyword arg `which='major'` to `autofmt_xdate`, which will in turn be passed to each call of `ax.get_xticklabels()`. The user could then select which='both', if wanted.

**Matplotlib version**
------------------
commit: None
python: 2.7.10.final.0
python-bits: 64
OS: Windows
OS-release: 7
machine: AMD64
processor: Intel64 Family 6 Model 23 Stepping 10, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.19.2
setuptools: 27.1.2
matplotlib: 1.5.1
...
although using matplotlib 1.5.1, the issue persists in the current master

**Comment by phobson** [comment]
Thanks for the clean and reproducible report.

I think `autofmt_xdate()` did format both major and minor tick labels. It just did it in two different ways to reduce redundant information on the plot.

One option would be to set all of the ticks to be major ticks. Another would be to use custom formatters. [This example](http://matplotlib.org/examples/pylab_examples/date_demo_convert.html) shows how to customize the formatting of date ticks. You would need to expand this section:

```python
ax.xaxis.set_major_locator(DayLocator())
ax.xaxis.set_minor_locator(HourLocator(arange(0, 25, 6)))
ax.xaxis.set_major_formatter(DateFormatter('%Y-%m-%d'))
```

To include `ax.xaxis.set_minor_formatter(DateFormatter(your_minor_fmt))`

**Comment by phobson** [comment]
Oh wait -- I see what's going on with the rotation.

This is better reported to pandas and they do some very clever tricksiness to locate and format their xticks (it's adaptive with the frequency of the time series)

**Comment by gitporst** [comment]
Thank you for your reply. I'm afraid, the pandas people will most likely delegate the issue back as their plot method works fine and it is the `autofmt_xdate()` method that does not do what I intend it to do. As stated,  `autofmt_xdate()` calls `ax.get_xticklabels()` which by default yields only the majorticklabels ( 
`def get_xticklabels(self, minor=False, which=None):`).
As I would like to keep pandas' clever tricksiness (their formatters), the easist way for me was to override `autofmt_xdate()` and pass `which='both'`.

**Comment by phobson** [comment]
So clarify this for me: what do you actually want?

Do you just want to rotate the all of the labels the same way? Do you want all of the labels to have the same day-month-year format?

**Comment by gitporst** [comment]
I would like the autofmt_xdate method to rotate all the labels the same way.

**Comment by tacaswell** [comment]
Exact work:

 - add `which` kwarg to `figure.autofmt_xdate` to control which tick label get rotated
 - numpy-docify the method docstring
 - add a test

**Comment by patniharshit** [comment]
I would like to do this but I have a couple of questions. 
1. What do you mean by 'numpy-docify' the method docstring ?
2. How to test the passing of a kwarg ?

**Comment by phobson** [comment]
@patniharshit 
"numpydoc" is a way of formatting docstrings such that sphinx + the numpydoc package can parse them nicely into html, latex, etc.

Here's the [specification for numpydoc](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt)

Here's an example of a matplotlib function with a good numpydoc-ified docstring:
https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/axes/_axes.py#L1866

For your second question, my approach would be write an image comparison test for each possible value of the kwarg. [This test](https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/tests/test_axes.py#L1604) and the one below it test two possible values of the `whis` kwarg in the function that draws boxplots.

**Comment by tacaswell** [comment]
Instead of image comparison, it might be better to just check the angle of the tick labels.  It will be faster (does not require the figure to be drawn), keeps the repo-size down, and will keep it independent of freetype changes, style changes, etc.

**Comment by anntzer** [comment]
Closed by #8157.
